### PR TITLE
feat: t11 - slow php reporting page

### PR DESCRIPTION
**Pull request by rtest12**
## t11 - slow php reporting page
A typical scenario:

-    you have a web app
-    that app has a backend written in PHP
-    most of the backend works fast and good for end users < 1000ms
-    however there is a slow page /reports/slow, that takes 10m to execute, which causes nginx and php to timeout
-    implement such solution so that nginx/php-fpm do not timeout and get the report when 10m passed.

Implementation nuances:

-    your nginx and php-fpm should run under nginx/nginx user/group.
-    use sleep php func to simulate such report, https://www.php.net/manual/en/function.sleep.php
-    implement a custom location /reports/fast, that will work fast and doesn't require any timeouts, and will use www pool of the PHP-FPM
-    implement a custom location /reports/slow, that will open /tmp/nginx/reports/slow.php script, which will take 10m to execute
-    during the /reports/slow, you should not lose the browser connection to the server.
-    IMPORTANT: you can't change any values for your core application that runs inside www pool defined in the PHP-FPM, for example you can't change timeout via www pool, as that may substantially degrade all the other URLs that are working good and fast - so think how you can do it.

When you open PR - pls package your solution as zip file with bash script run.sh that should

-    create all necessary files and/or folders.
-    set proper permissions to created files/folders
-    start nginx in foreground mode (so all errors, logs are printed on `stdout/stderr`)
-    start php-fpm in foreground mode (so all errors, logs are printed on `stdout/stderr`)
-    `curl https://localhost/reports/fast` should output "FAST REPORT" via PHP script
-    `curl https://localhost/reports/slow` should output "SLOW REPORT" via PHP script with 10m delay without losing any connection (i.e. no timeout or upstream timeout error in the nginx)

## My output log
[t11_log.txt](https://github.com/rtest12/devops-camp-2023/files/11268436/t11_log.txt)

### CODE REVIEW

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
zip file had to be attached to summary
all other files had to be added as is into PR/git remote so we can review them.

**Comment by dmitry-mightydevops** [comment]
Also in summary pls include `time` calls, i.e.
```
$ time curl https://camp-php.local/reports/fast
FAST REPORT

________________________________________________________
Executed in   52.49 millis    fish           external

$ time curl https://camp-php.local/reports/slow
SLOW REPORT

________________________________________________________
Executed in  600.05 secs      fish           external
```

**Comment by rtest12** [comment]
> zip file had to be attached to summary all other files had to be added as is into PR/git remote so we can review them.

@dmitry-mightydevops 
Ok, added everything.

File: chapter2/tests/t11/Task_11.zip
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/Task_11.zip:1]: binary files should not be in the remote

File: chapter2/tests/t11/fastcgi_params [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+fastcgi_param  GATEWAY_INTERFACE  CGI/1.1;
+fastcgi_param  SERVER_SOFTWARE    nginx/$nginx_version;
+
+fastcgi_param  REMOTE_ADDR        $remote_addr;
+fastcgi_param  REMOTE_PORT        $remote_port;
+fastcgi_param  REMOTE_USER        $remote_user;
+fastcgi_param  SERVER_ADDR        $server_addr;
+fastcgi_param  SERVER_PORT        $server_port;
+fastcgi_param  SERVER_NAME        $server_name;
+fastcgi_param  SCRIPT_FILENAME    $document_root$fastcgi_script_name;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/fastcgi_params]: prefer real path name, I have mentioned this in the lecture.

File: chapter2/tests/t11/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+
+	upstream php-fpm-slow {
+		server unix:/run/php/php8.1-fpm-slow.sock fail_timeout=0;
+	}
+
+
+	upstream php-fpm {
+		server unix:/run/php/php8.1-fpm.sock;
+	}
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/nginx.conf]: prefer upstreams before servers/locations

File: chapter2/tests/t11/nginx.conf [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+	ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
+	ssl_prefer_server_ciphers on;
+
+
+	error_log /dev/stderr;
+	access_log /dev/stdout;
+
+
+	gzip on;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/nginx.conf]: that should go on top of the file. Your servers/locations should be at the bottom

File: chapter2/tests/t11/pool.d/slowww.conf [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+; Depth of slow log stack trace.
+; Default Value: 20
+;request_slowlog_trace_depth = 20
+
+; The timeout for serving a single request after which the worker process will
+; be killed. This option should be used when the 'max_execution_time' ini option
+; does not stop script execution for some reason. A value of '0' means 'off'.
+; Available units: s(econds)(default), m(inutes), h(ours), or d(ays)
+; Default Value: 0
+request_terminate_timeout = 600
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/pool.d/slowww.conf]: you don't need it. It has nothing to do with the "timeout" solution we're after - it basically kill the underlying worker after 10m. So comment it and rely on max_execution_time, which you should set in the pool as php_admin value

File: chapter2/tests/t11/pool.d/slowww.conf [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+; (pool, global or /usr)
+
+; Default Value: nothing is defined by default except the values in php.ini and
+;                specified at startup with the -d argument
+;php_admin_value[sendmail_path] = /usr/sbin/sendmail -t -i -f www@my.domain.com
+;php_flag[display_errors] = off
+;php_admin_value[error_log] = /var/log/fpm-php.www.log
+ php_admin_value[error_log] = /proc/self/fd/2
+;php_admin_flag[log_errors] = on
+;php_admin_value[memory_limit] = 32M
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/pool.d/slowww.conf]: you should set `max_execution_time` explicitly in the pool
https://code.tutsplus.com/tutorials/how-to-increase-max_execution_time-in-php--cms-37017

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter2/tests/t11/Task_11.zip
Comment (rtest12) [reply chapter2/tests/t11/Task_11.zip:1]: @dmitry-mightydevops 
All fixed

**Comment by dmitry-mightydevops** [comment]
@rtest12 pls follow the workflow - every single comment in the feedback should be addressed and then replied either with question or "Done".

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
not addressed.

**Comment by rtest12** [comment]
> @rtest12 pls follow the workflow - every single comment in the feedback should be addressed and then replied either with question or "Done".

@dmitry-mightydevops 
I did not know, apparently I missed something, ok, it will be done!

File: chapter2/tests/t11/pool.d/slowww.conf (hunk 5 above)
Comment (rtest12) [reply chapter2/tests/t11/pool.d/slowww.conf]: @dmitry-mightydevops  
Done

File: chapter2/tests/t11/pool.d/slowww.conf (hunk 4 above)
Comment (rtest12) [reply chapter2/tests/t11/pool.d/slowww.conf]: > you don't need it. It has nothing to do with the "timeout" solution we're after - it basically kill the underlying worker after 10m. So comment it and rely on max_execution_time, which you should set in the pool as php_admin value

@dmitry-mightydevops 
Ok, done

File: chapter2/tests/t11/nginx.conf (hunk 3 above)
Comment (rtest12) [reply chapter2/tests/t11/nginx.conf]: > that should go on top of the file. Your servers/locations should be at the bottom

@dmitry-mightydevops 
Done

File: chapter2/tests/t11/nginx.conf (hunk 2 above)
Comment (rtest12) [reply chapter2/tests/t11/nginx.conf]: > prefer upstreams before servers/locations

@dmitry-mightydevops 
Done

File: chapter2/tests/t11/fastcgi_params (hunk 1 above)
Comment (rtest12) [reply chapter2/tests/t11/fastcgi_params]: > prefer real path name, I have mentioned this in the lecture.

@dmitry-mightydevops 
Like so?

**Comment by rtest12** [comment]
> not addressed.
@dmitry-mightydevops 
Ok, i got)

File: chapter2/tests/t11/fastcgi_params (hunk 1 above)
Comment (dmitry-mightydevops) [reply chapter2/tests/t11/fastcgi_params]: no, that way

```
fastcgi_param  SCRIPT_FILENAME    $realpath_root$fastcgi_script_name;
```

http://nginx.org/en/docs/http/ngx_http_core_module.html

**$realpath_root**
an absolute pathname corresponding to the [root](http://nginx.org/en/docs/http/ngx_http_core_module.html#root) or [alias](http://nginx.org/en/docs/http/ngx_http_core_module.html#alias) directive’s value for the current request, with all symbolic links resolved to real paths

**$document_root**
[root](http://nginx.org/en/docs/http/ngx_http_core_module.html#root) or [alias](http://nginx.org/en/docs/http/ngx_http_core_module.html#alias) directive’s value for the current request

File: chapter2/tests/t11/nginx.conf [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+	types_hash_max_size 4096;
+	include /etc/nginx/mime.types;
+	default_type application/octet-stream;
+	ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3;
+	ssl_prefer_server_ciphers on;
+	error_log /dev/stderr;
+	access_log /dev/stdout;
+
+
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/nginx.conf]: too many spaces

File: chapter2/tests/t11/pool.d/slowww.conf [hunk 7]
```diff
@@ -0,0 +1,466 @@
+; Start a new pool named 'www'.
+; the variable $pool can be used in any directive and will be replaced by the
+; pool name ('www' here)
+[slowww]
+
+; Also added for the correct execution of the slow report
+php_admin_value[max_execution_time] = 605
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/pool.d/slowww.conf]: prefer to keep every `php_*` directive at the bottom of the page.

Comment (dmitry-mightydevops) [reply chapter2/tests/t11/pool.d/slowww.conf]: remove all other options you don't need (or defaults). So it's easier to see where slow.conf is different vs www.conf

File: chapter2/tests/t11/nginx.conf [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+			root /tmp/camp/reports/fast;
+			fastcgi_pass php-fpm;
+			try_files /fast.php =404;
+		}
+
+		location /reports/slow/ {
+			root /tmp/camp/reports/slow;
+			fastcgi_pass php-fpm-slow;
+			fastcgi_read_timeout 605;
+			try_files /slow.php =404;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/nginx.conf]: define root at the server level, in your case it should be `/tmp/camp`
then have try_files as the first line in the location context. all `fastcgi_*` directives should be at the end of the location context.

File: chapter2/tests/t11/pool.d/slowww.conf [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+
+; Per pool prefix
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/pool.d/slowww.conf]: all default values should be defined globally in php-fpm.conf and `concrete` pools should only define the "DIFF" that way you can clearly see what's happening in each pool. Right now it's hard.

File: chapter2/tests/t11/pool.d/www.conf [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+; (pool, global or /usr)
+
+; Default Value: nothing is defined by default except the values in php.ini and
+;                specified at startup with the -d argument
+;php_admin_value[sendmail_path] = /usr/sbin/sendmail -t -i -f www@my.domain.com
+;php_flag[display_errors] = off
+ php_admin_value[error_log] = /proc/self/fd/2
+;php_admin_value[error_log] = /var/log/fpm-php.www.log
+ php_admin_flag[log_errors] = on
+;php_admin_value[memory_limit] = 32M
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/pool.d/www.conf]: you www.pool should also define the timeout and make sure no scripts can be running above 60s. Same for the related nginx location ctx.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter2/tests/t11/fastcgi_params (hunk 1 above)
Comment (rtest12) [reply chapter2/tests/t11/fastcgi_params]: > no, that way
> 
> ```
> fastcgi_param  SCRIPT_FILENAME    $realpath_root$fastcgi_script_name;
> ```
> 
> http://nginx.org/en/docs/http/ngx_http_core_module.html
> 
> **$realpath_root** an absolute pathname corresponding to the [root](http://nginx.org/en/docs/http/ngx_http_core_module.html#root) or [alias](http://nginx.org/en/docs/http/ngx_http_core_module.html#alias) directive’s value for the current request, with all symbolic links resolved to real paths
> 
> **$document_root** [root](http://nginx.org/en/docs/http/ngx_http_core_module.html#root) or [alias](http://nginx.org/en/docs/http/ngx_http_core_module.html#alias) directive’s value for the current request

@dmitry-mightydevops 
Done

File: chapter2/tests/t11/nginx.conf (hunk 6 above)
Comment (rtest12) [reply chapter2/tests/t11/nginx.conf]: > too many spaces

@dmitry-mightydevops 
Done

File: chapter2/tests/t11/pool.d/slowww.conf (hunk 7 above)
Comment (rtest12) [reply chapter2/tests/t11/pool.d/slowww.conf]: > prefer to keep every `php_*` directive at the bottom of the page.

@dmitry-mightydevops 
Done

Comment (rtest12) [reply chapter2/tests/t11/pool.d/slowww.conf]: > remove all other options you don't need (or defaults). So it's easier to see where slow.conf is different vs [www.conf](http://www.conf)

@dmitry-mightydevops 
Done

File: chapter2/tests/t11/nginx.conf (hunk 8 above)
Comment (rtest12) [reply chapter2/tests/t11/nginx.conf]: > define root at the server level, in your case it should be `/tmp/camp` then have try_files as the first line in the location context. all `fastcgi_*` directives should be at the end of the location context.

@dmitry-mightydevops 
Done

File: chapter2/tests/t11/pool.d/slowww.conf (hunk 9 above)
Comment (rtest12) [reply chapter2/tests/t11/pool.d/slowww.conf]: > all default values should be defined globally in php-fpm.conf and `concrete` pools should only define the "DIFF" that way you can clearly see what's happening in each pool. Right now it's hard.

@dmitry-mightydevops 
Done

File: chapter2/tests/t11/pool.d/www.conf (hunk 10 above)
Comment (rtest12) [reply chapter2/tests/t11/pool.d/www.conf]: > you [www.pool](http://www.pool) should also define the timeout and make sure no scripts can be running above 60s. Same for the related nginx location ctx.

@dmitry-mightydevops 
Done

File: chapter2/tests/t11/nginx.conf [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+events {
+	worker_connections 768;
+}
+
+http {
+
+	types_hash_max_size 4096;
+	include /etc/nginx/mime.types;
+	default_type application/octet-stream;
+	ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t11/nginx.conf]: prefer 1.2 and 1.3 only.

**Review by dmitry-mightydevops — APPROVED** [review]

File: chapter2/tests/t11/nginx.conf (hunk 11 above)
Comment (rtest12) [reply chapter2/tests/t11/nginx.conf]: > prefer 1.2 and 1.3 only.

@dmitry-mightydevops 
Ok, done

**Pull request merged**